File: src/main/java/bot/boobbot/BoobBot.java
package bot.boobbot;

import bot.boobbot.flight.Command;
import bot.boobbot.handlers.EventHandler;
import bot.boobbot.handlers.MessageHandler;
import bot.boobbot.misc.Constants;
import bot.boobbot.misc.EventWaiter;
import bot.boobbot.misc.RequestUtil;
import ch.qos.logback.classic.Level;
import ch.qos.logback.classic.Logger;
import io.sentry.Sentry;
import net.dv8tion.jda.bot.sharding.DefaultShardManagerBuilder;
import net.dv8tion.jda.bot.sharding.ShardManager;
import net.dv8tion.jda.core.JDAInfo;
import net.dv8tion.jda.core.entities.Game;
import org.reflections.Reflections;
import org.slf4j.LoggerFactory;

import java.lang.reflect.Modifier;
import java.util.HashMap;
import java.util.Map;

import static bot.boobbot.misc.Constants.SENTRY_DSN;


public class BoobBot {

    public static Logger log = (Logger) LoggerFactory.getLogger(BoobBot.class);
    private static boolean isDebug = false;
    private static ShardManager shardManager;
    private static final Map<String, Command> commands = new HashMap<>();
    private static final EventWaiter waiter = new EventWaiter();
    private static final RequestUtil requestUtil = new RequestUtil();


    public static void main(String[] args) throws Exception {
        Sentry.init(SENTRY_DSN);
        log.info("--- BoobBot.jda ---");
        log.info(JDAInfo.VERSION);

        isDebug = args.length > 0 && args[0].contains("debug");
        String token = isDebug ? Constants.DEBUG_TOKEN : Constants.TOKEN;

        if (isDebug) {
            log.warn("Running in debug mode");
            log.setLevel(Level.DEBUG);
        }

        shardManager = new DefaultShardManagerBuilder()
                .setGame(Game.playing("bbhelp | bbinvite"))
                .addEventListeners(new MessageHandler(), new EventHandler(), waiter)
                .setToken(token)
                .setShardsTotal(-1)
                .build();
    }

    private static void loadCommands() {
        Reflections reflections = new Reflections("bot.boobbot.commands");

        reflections.getSubTypesOf(Command.class).forEach(command -> {
            if (Modifier.isAbstract(command.getModifiers()) || command.isInterface()) {
                return;
            }

            try {
                Command cmd = command.newInstance();

                if (cmd.getProperties() == null) {
                    log.warn("Command `" + cmd.getName() + "` is missing CommandProperties annotation. Will not load.");
                    return;
                }

                commands.put(cmd.getName(), cmd);
            } catch (InstantiationException | IllegalAccessException e) {
                log.error("Failed to load command `" + command.getSimpleName() + "`", e);
            }
        });

        log.info("Successfully loaded " + commands.size() + " commands!");
    }

    public static ShardManager getShardManager() {
        return shardManager;
    }

    public static Map<String, Command> getCommands() {
        return commands;
    }

    public static boolean isDebug() {
        return isDebug;
    }

    public static EventWaiter getWaiter() {
        return waiter;
    }

    public static RequestUtil getRequestUtil() {
        return requestUtil;
    }

}

File: src/main/java/bot/boobbot/BoobBot.kt
package bot.boobbot

import bot.boobbot.flight.Command
import bot.boobbot.handlers.EventHandler
import bot.boobbot.handlers.MessageHandler
import bot.boobbot.misc.Constants
import bot.boobbot.misc.EventWaiter
import bot.boobbot.misc.Formats.Companion.getReadyFormat
import bot.boobbot.misc.GuildMusicManager
import bot.boobbot.misc.RequestUtil
import ch.qos.logback.classic.Level
import ch.qos.logback.classic.Logger
import com.sedmelluq.discord.lavaplayer.jdaudp.NativeAudioSendFactory
import com.sedmelluq.discord.lavaplayer.player.AudioConfiguration
import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager
import com.sedmelluq.discord.lavaplayer.player.DefaultAudioPlayerManager
import com.sedmelluq.discord.lavaplayer.source.AudioSourceManagers
import com.sedmelluq.discord.lavaplayer.source.youtube.YoutubeAudioSourceManager
import io.ktor.server.netty.*
import io.ktor.routing.*
import io.ktor.application.*
import io.ktor.http.*
import io.ktor.response.*
import io.ktor.server.engine.*
import io.sentry.Sentry
import net.dv8tion.jda.bot.sharding.DefaultShardManagerBuilder
import net.dv8tion.jda.bot.sharding.ShardManager
import net.dv8tion.jda.core.JDAInfo
import net.dv8tion.jda.core.entities.Game
import net.dv8tion.jda.core.entities.Guild
import net.dv8tion.jda.core.events.ReadyEvent
import net.dv8tion.jda.core.hooks.ListenerAdapter
import org.reflections.Reflections
import org.slf4j.LoggerFactory
import java.lang.reflect.Modifier
import java.util.*
import java.util.concurrent.ConcurrentHashMap


class BoobBot : ListenerAdapter() {

    override fun onReady(event: ReadyEvent) {
        if (shardManager.shardsRunning == shardManager.shardsTotal && !isReady) {
            isReady = true
            // health check for status page
            embeddedServer(Netty, 8008) {
                routing {
                    get("/health") {
                        call.respondText("{health: ok, ping: ${shardManager.averagePing}}", ContentType.Application.Json)
                    }
                }
            }.start(wait = false)
            log.info(getReadyFormat(event.jda, home))
        }
    }

    companion object {
        lateinit var playerManager: AudioPlayerManager
        lateinit var musicManagers: Map<String, GuildMusicManager>

        val log = LoggerFactory.getLogger(BoobBot::class.java) as Logger
        val startTime = System.currentTimeMillis()

        var isDebug = false
            private set

        lateinit var shardManager: ShardManager
            private set

        var isReady = false
            private set

        private val commands = HashMap<String, Command>()
        val waiter = EventWaiter()
        val requestUtil = RequestUtil()

        val home: Guild?
            get() = shardManager.getGuildById(Constants.HOME_GUILD)


        @Throws(Exception::class)
        @JvmStatic
        fun main(args: Array<String>) {
            Sentry.init(Constants.SENTRY_DSN)
            playerManager = DefaultAudioPlayerManager()
            AudioSourceManagers.registerRemoteSources(playerManager)
            playerManager.registerSourceManager(YoutubeAudioSourceManager())
           // playerManager.registerSourceManager(PornHubAudioSourceManager()) //TODO add this stuff
            //playerManager.registerSourceManager(RedTubeAudioSourceManager())
            playerManager.configuration.opusEncodingQuality = 10
            playerManager
                    .configuration.resamplingQuality = AudioConfiguration.ResamplingQuality.HIGH

            musicManagers = ConcurrentHashMap()
            log.info("--- BoobBot.jda ---")
            log.info(JDAInfo.VERSION)

            isDebug = args.isNotEmpty() && args[0].contains("debug")
            val token = if (isDebug) Constants.DEBUG_TOKEN else Constants.TOKEN

            if (isDebug) {
                log.warn("Running in debug mode")
                log.level = Level.DEBUG
            }

            shardManager = DefaultShardManagerBuilder()
                    .setGame(Game.playing("bbhelp | bbinvite"))
                    .setAudioSendFactory(NativeAudioSendFactory())
                    .addEventListeners(BoobBot(), MessageHandler(), EventHandler(), waiter)
                    .setToken(token)
                    .setShardsTotal(-1)
                    .build()

            loadCommands()
        }

        private fun loadCommands() {
            val reflections = Reflections("bot.boobbot.commands")

            reflections.getSubTypesOf(Command::class.java).forEach { command ->
                if (Modifier.isAbstract(command.modifiers) || command.isInterface) {
                    return@forEach
                }

                try {
                    val cmd = command.getDeclaredConstructor().newInstance()
                    if (!cmd.hasProperties) {
                        return@forEach log.warn("Command `${cmd.name}` is missing CommandProperties annotation. Will not load.")
                    }

                    commands[cmd.name] = cmd
                } catch (e: InstantiationException) {
                    log.error("Failed to load command `${command.simpleName}`", e)
                } catch (e: IllegalAccessException) {
                    log.error("Failed to load command `${command.simpleName}`", e)
                }
            }

            log.info("Successfully loaded " + commands.size + " commands!")
            // TODO: Eval
        }

        fun getCommands(): Map<String, Command> {
            return commands
        }

    }

}
